src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -126,30 +119,99 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
    * @param thisNode the input data Node object
    * @return a useful node used to construct the Guava Graph
    */
-  private GraphNode protoNodeToGraphNode(Node thisNode) {
-    return GraphNode.create(thisNode.getName(), thisNode.getTokenList(), thisNode.getMetadata());
+  public GraphNode protoNodeToGraphNode(Node thisNode) {
+    List<String> newTokenList = new ArrayList<>();
+    newTokenList.addAll(thisNode.getTokenList());
+    Struct newMetadata = Struct.newBuilder().mergeFrom(thisNode.getMetadata()).build();
+    return GraphNode.create(thisNode.getName(), newTokenList, newMetadata);
+  }
+
+  /*
+   * Changes the graph according to the given mutation object
+   * @param mut the mutation to affect
+   * @param graph the Guava graph to mutate
+   * @param graphNodesMap a reference of existing nodes, also to be mutated
+   * @return true if the mutation was successful, false otherwise
+   */
+  public boolean mutateGraph(
+      Mutation mut, MutableGraph<GraphNode> graph, Map<String, GraphNode> graphNodesMap) {
+    // Nodes affected by the mutation
+    // second node only applicable for adding an edge and removing an edge
+    String startName = mut.getStartNode();
+    String endName = mut.getEndNode();
+
+    // Getting the corresponding graph nodes from the graph map
+    GraphNode startNode = graphNodesMap.get(startName);
+    GraphNode endNode = graphNodesMap.get(endName);
+
+    switch (mut.getType()) {
+      case ADD_NODE:
+        if (graphNodesMap.containsKey(startName)) {
+          // adding a duplicate node
+          return false;
+        }
+        // Create a new node with the given name and add it to the graph and the map
+        GraphNode newGraphNode =
+            GraphNode.create(startName, new ArrayList<>(), Struct.newBuilder().build());
+        graph.addNode(newGraphNode);
+        graphNodesMap.put(startName, newGraphNode);
+        break;
+      case ADD_EDGE:
+        if (startNode != null && endNode != null) { // Check nodes exist before adding an edge
+          graph.putEdge(startNode, endNode);
+        } else {
+          return false;
+        }
+        break;
+      case DELETE_NODE:
+        if (startNode != null) { // Check node exists before removing
+          graph.removeNode(startNode); // This will remove all edges associated with startNode
+          graphNodesMap.remove(startName);
+        } else {
+          return false;
+        }
+        break;
+      case DELETE_EDGE:
+        if (startNode != null && endNode != null) { // Check nodes exist before removing edge
+          graph.removeEdge(startNode, endNode);
+        } else {
+          return false;
+        }
+        break;
+      case CHANGE_TOKEN:
+        if (startNode != null) {
```

Comment:
Here you'll be returning true if startNode is null?
I think you can change it to:
```
if (startNode == null) {
     return false;
}
return changeNodeToken(...) // since it's already a boolean.
```

Reply:
I thought it would fall through and return false from the default because there's no break but to be consistent with the others I'll change it to that :)

Reply:
Ah I missed that there was no break :)